doc/developer-guide/internal-libraries/MemArena.en.rst

```diff
@@ -14,223 +14,137 @@
+heuristic on whether :class:`MemArena` is an appropriate choice.
+
+While :class:`MemArena` will normally allocate memory from an internal chunk, if the allocation
+request is large (more than a memory page) and there is not enough space in the current internal
+block, a block just for that allocation will be created. This is useful if the purpose of
+:class:`MemArena` is to track blocks of memory more than reduce the number of system level
+allocations.
 
 Reference
 +++++++++
 
 .. class:: MemArena
 
-   .. class:: Block
-      
-      Underlying memory allocated is owned by the :class:`Block`. A linked list. 
-
-      .. member:: size_t size
-      .. member:: size_t allocated
-      .. member:: std::shared_ptr<Block> next
-      .. function:: Block(size_t n)
-      .. function:: char* data()
-
    .. function:: MemArena()
 
       Construct an empty arena.
 
-   .. function:: explicit MemArena(size_t n)
+   .. function:: MemArena(size_t n)
 
-      Construct an arena with :arg:`n` bytes. 
+      Construct an arena with an initial internal block with at least :arg:`n` bytes of free space.
 
    .. function:: MemSpan alloc(size_t n)
 
-      Allocate an :arg:`n` byte chunk of memory in the arena.
-
-   .. function:: MemArena& freeze(size_t n = 0)
+      Allocate memory of size :arg:`n` bytes in the arena.
 
-      Block all further allocation from any existing internal blocks. If :arg:`n` is zero then on the next allocation request a block twice as large as the current generation, otherwise the next internal block will be large enough to hold :arg:`n` bytes.
+   .. function:: template < typename T, typename ... Args > T * make(Args&& ... args)
 
-   .. function:: MemArena& thaw()
+      Create an instance of :arg:`T`. :code:`sizeof(T)` bytes of memory are allocated from the arena
+      and the constructor invoked. This method takes any set of arguments, which are passed to
+      the constructor. A pointer to the newly constructed instance of :arg:`T` is returned. Note if
+      the instance allocates other memory that memory will not be in the arena. Example constructing
+      a :code:`std::string_view` ::
 
-      Unallocate all internal blocks that were allocated before the current generation. 
-    
-   .. function:: MemArena& empty()
-     
-      Empties the entire arena and deallocates all underlying memory. Next block size will be equal to the sum of all allocations before the call to empty.
+         std::string_view * sv = arena.make<std::string_view>(pointer, n);
 
-   .. function:: size_t size() const 
+   .. function:: MemArena& freeze(bool flag)
 
-      Get the current generation size. The default size of the arena is 32KB unless otherwise specified. 
+      If :arg:`flag` is :code:`true`, stop allocating from existing internal memory blocks. These
+      blocks are now "frozen". Adjust the size of the next allocation block, if necessary, to be
```

Comment:
Why does freezing and un-freezing need to allocate or deallocate memory?  Shouldn't that have been done with reserve and deleting the object?

Reply:
Freezing doesn't allocate memory. Unfreezing cleans up the old memory. If it didn't do that, what would be the point of freezing at all? In theory it could be done by creating a new instance of `MemArena` and copying in to that, but that requires two allocations (one for the `MemArena`, and another for the new internal block) instead of just one, in addition to making the containing object construction more complex (i.e. it's no longer possible to declare the member simply as `MemArena`). Beyond that, one goal was to emulate the behavior of `HdrHeap` and that is how it works.